node/network/statement-distribution/src/vstaging/grid.rs

```diff
@@ -0,0 +1,2247 @@
+}
+
+// A utility struct for keeping track of metadata about candidates
+// we have confirmed as having been backed.
+#[derive(Debug, Clone)]
+struct KnownBackedCandidate {
+	group_index: GroupIndex,
+	local_knowledge: StatementFilter,
+	mutual_knowledge: HashMap<ValidatorIndex, MutualKnowledge>,
+}
+
+impl KnownBackedCandidate {
+	fn has_received_manifest_from(&self, validator: ValidatorIndex) -> bool {
+		self.mutual_knowledge
+			.get(&validator)
+			.map_or(false, |k| k.remote_knowledge.is_some())
+	}
+
+	fn has_sent_manifest_to(&self, validator: ValidatorIndex) -> bool {
+		self.mutual_knowledge
+			.get(&validator)
+			.map_or(false, |k| k.local_knowledge.is_some())
+	}
+
+	fn manifest_sent_to(&mut self, validator: ValidatorIndex, local_knowledge: StatementFilter) {
+		let k = self
+			.mutual_knowledge
+			.entry(validator)
+			.or_insert_with(|| MutualKnowledge { remote_knowledge: None, local_knowledge: None });
+
+		k.local_knowledge = Some(local_knowledge);
+	}
+
+	fn manifest_received_from(
+		&mut self,
+		validator: ValidatorIndex,
+		remote_knowledge: StatementFilter,
+	) {
+		let k = self
+			.mutual_knowledge
+			.entry(validator)
+			.or_insert_with(|| MutualKnowledge { remote_knowledge: None, local_knowledge: None });
+
+		k.remote_knowledge = Some(remote_knowledge);
+	}
+
+	fn direct_statement_senders(
+		&self,
+		group_index: GroupIndex,
+		originator_index_in_group: usize,
+		statement_kind: StatementKind,
+	) -> Vec<ValidatorIndex> {
+		if group_index != self.group_index {
+			return Vec::new()
+		}
+
+		self.mutual_knowledge
+			.iter()
+			.filter(|(_, k)| k.remote_knowledge.is_some())
+			.filter(|(_, k)| {
+				k.local_knowledge
+					.as_ref()
+					.map_or(false, |r| !r.contains(originator_index_in_group, statement_kind))
+			})
+			.map(|(v, _)| *v)
+			.collect()
+	}
+
+	fn direct_statement_recipients(
+		&self,
+		group_index: GroupIndex,
+		originator_index_in_group: usize,
+		statement_kind: StatementKind,
+	) -> Vec<ValidatorIndex> {
+		if group_index != self.group_index {
+			return Vec::new()
+		}
+
+		self.mutual_knowledge
+			.iter()
+			.filter(|(_, k)| k.local_knowledge.is_some())
+			.filter(|(_, k)| {
+				k.remote_knowledge
+					.as_ref()
+					.map_or(false, |r| !r.contains(originator_index_in_group, statement_kind))
+			})
+			.map(|(v, _)| *v)
+			.collect()
+	}
+
+	fn note_fresh_statement(
+		&mut self,
+		statement_index_in_group: usize,
+		statement_kind: StatementKind,
+	) -> bool {
+		let really_fresh = !self.local_knowledge.contains(statement_index_in_group, statement_kind);
+		self.local_knowledge.set(statement_index_in_group, statement_kind);
+
+		really_fresh
+	}
+
+	fn sent_or_received_direct_statement(
+		&mut self,
+		validator: ValidatorIndex,
+		statement_index_in_group: usize,
+		statement_kind: StatementKind,
+	) {
+		if let Some(k) = self.mutual_knowledge.get_mut(&validator) {
+			if let (Some(r), Some(l)) = (k.remote_knowledge.as_mut(), k.local_knowledge.as_mut()) {
```

Comment:
Both need to be some?

Reply:
yes, because in the grid mode we are only allowed to send direct statements after a manifest exchange. Both 'remote_knowledge' and 'local_knowledge' being Some is a check for that.